Assets/Scripts/Layers/Adapters/DataTypeChain.cs

```diff
@@ -94,22 +92,28 @@ private IEnumerator DownloadAndCheckSupport(string url)
         /// <returns></returns>
         private IEnumerator DownloadDataToLocalCache(LocalFile urlAndData)
         {
-            var url = urlAndData.SourceUrl;
-            var uwr = UnityWebRequest.Get(url);
-            var optionalExtention = Path.GetExtension(url).Split("?")[0];
-            var guidFilename = Guid.NewGuid().ToString() + optionalExtention;
-            string path = Path.Combine(Application.persistentDataPath, guidFilename);
-
-            uwr.downloadHandler = new DownloadHandlerFile(path);
-            yield return uwr.SendWebRequest();
-            if (uwr.result == UnityWebRequest.Result.Success)
-            {
-                urlAndData.LocalFilePath = path;
-            }
-            else
+            var futureFileInfo = Resource<FileInfo>.At(urlAndData.SourceUrl).Value;
+            
+            // We want to use and manipulate urlAndData, so we 'curry' it by wrapping a method call in a lambda 
+            futureFileInfo.Then(info => DownloadSucceeded(urlAndData, info));
+            futureFileInfo.Catch(error => DownloadFailed(urlAndData, error));
+            
+            yield return Uxios.WaitForRequest(futureFileInfo);
```

Comment:
Usually, you need not wait for a Promise to complete; the only reason this is here is because this IEnumerator is used in a chain of IEnumerators starting in another coroutine, and continuing after this one. Without refactoring that further, a yield instruction here is necessary. (Just as you do not need to put an Uxios call in a coroutine; Uxios handles the async stuff all by itself)

Reply:
will putting yield return null; here work too then? just for my curiosity

Reply:
Yield return null will skip 1 frame, but we need to wait for the async process that is in the promise to complete (I wanted to call the method 'WaitForTheFuture' because another name for promises is 'Futures', but that might have been too much tongue in cheek ;) ).

What the WaitForRequest yield instruction does is similar to a `yield return null` every frame until the Promise has reached state "Resolved" (i.e. Done). It is similar to what you would do with UnityWebRequest when instead of yielding SendRequest that you enter a loop and `yield return null` until the UnityWebRequest returns a completed state.

So in short: 1 `yield return null` wouldn't work because you need to `yield return null` until the Promise is done; and that is what this does :)